### ISSUE TITLE
termux-notification-list hangs: Calling startActivity() from outside of an Activity

### ISSUE THREAD

**Issue by AgentConDier** [termux/termux-api#286]
**Problem description**
termux-notification-list hangs indefinitely and does not output anything. Seems to be caused by `android.util.AndroidRuntimeException: Calling startActivity() from outside of an Activity  context requires the FLAG_ACTIVITY_NEW_TASK flag. Is this really what you want?` (see adb log)

**Steps to reproduce**
Run `termux-notification-list`

**Expected behavior**
Notification list printed to stdout in json format

**Additional information**
<details><summary>termux-info</summary><pre>Updatable packages:<br>mosh/stable 1.3.2-16 aarch64 [upgradable from: 1.3.2-16]<br>Subscribed repositories:<br>https://dl.bintray.com/grimler/science-packages-24 science/stable<br>https://dl.bintray.com/grimler/game-packages-24 games/stable<br>https://dl.bintray.com/termux/termux-packages-24 stable/main<br>System information:<br>Linux localhost 4.9.148 #1 SMP PREEMPT Thu Jul 4 00:12:42 CST 2019 aarch64 Android<br>Termux-packages arch:<br>aarch64<br>Android version:<br>9<br>Device manufacturer:<br>HUAWEI<br>Device model:<br>LYA-L29</pre></details><details><summary>adb logcat</summary><pre>08-23 14:49:43.487 12625 12625 I appproc : CLASSPATH=/data/data/com.termux/files/usr/libexec/termux-am/am.apk<br>08-23 14:49:43.487 12625 12625 I appproc : Command=/system/bin/app_process / com.example.termuxam.Am broadcast --user 0 -n com.termux.api/.TermuxApiReceiver --es socket_input ea592991c5d68171-3150-4f5c-8f46-b569a7714eda7000c53c0bb3 --es socket_output 2c8e5142b3e943fd-3150-411e-ac0e-96de4318ef75ef6631dc424e --es api_method NotificationList<br>08-23 14:49:43.487 12625 12625 D AndroidRuntime: >>>>>> START com.android.internal.os.RuntimeInit uid 10224 <<<<<<<br>08-23 14:49:43.494 12625 12625 I AndroidRuntime: readDownloadBoosterConfig: 'false'<br>08-23 14:49:43.550   867   867 D SensorHub: -------------------------->SENSOR_TYPE_PROXIMITY data[0] = 5.000000<br>08-23 14:49:43.629   867   867 D SensorHub: -------------------------->SENSOR_TYPE_PROXIMITY data[0] = 5.000000<br>08-23 14:49:43.646  2179  2179 D HwInnerWifiManagerImpl: fetchWifiSignalInfoForVoWiFi<br>08-23 14:49:43.656  1262  4029 E WificondControl: Noise: -92, Snr: 65, Chload: 53<br>08-23 14:49:43.658  1262  4029 D HwWifiStateMachine: rssi=-60,nativeTxBad=157, nativeTxGood=962919, dpktcnt=0, linkSpeed=78, frequency=2412, noise=0, mac=c80e14<br>08-23 14:49:43.659  2179  2802 W VoWifiQoe_N: processSignalInfoInd phoneId 1 is not qoe started!<br>08-23 14:49:43.662 12625 12625 I /system/bin/app_process: Reinit property: dalvik.vm.checkjni= false<br>08-23 14:49:43.666 12625 12625 V HwFrameworkFactory: : successes to get AllImpl object and return....<br>08-23 14:49:43.679 12625 12625 I HWSERVICES: hwnative_get_component_register:libdrmbitmap,JNI<br>08-23 14:49:43.699 12625 12625 E Zygote  : register_com_android_internal_os_Zygote num = 0<br>08-23 14:49:43.700 12625 12625 I         : power log dlsym ok<br>08-23 14:49:43.702 12625 12625 D         : APS:IFLoad: importExternalFunctions, dlopen libhwaps ok<br>08-23 14:49:43.705 12625 12625 E libc    : Access denied finding property "ro.config.vol_steps"<br>08-23 14:49:43.709   867   867 D SensorHub: -------------------------->SENSOR_TYPE_PROXIMITY data[0] = 5.000000<br>08-23 14:49:43.719 12625 12625 D AndroidRuntime: Calling main entry com.example.termuxam.Am<br>08-23 14:49:43.724   803 20323 I iked    : [INFO]: ike_localsocket.c:2851: Get WakeLock begin. vowifi_wake_lock_network<br>08-23 14:49:43.724   803 20323 D iked    : [DEBUG]: sockmisc.c:201: memcpy to = 192.***.***.24[4500] ,len=16,sslen=16,tolen=128<br>08-23 14:49:43.724   803 20323 D iked    : [DEBUG]: isakmp.c:863: isakmp_handler socket =33 ,local: 192.***.***.24[4500] remote 89.***.***.64[4500]<br>08-23 14:49:43.724   803 20323 D iked    : [DEBUG]: sockmisc.c:201: memcpy to = 192.***.***.24[4500] ,len=0,sslen=16,tolen=0<br>08-23 14:49:43.724   803 20323 D iked    : [DEBUG]: isakmp.c:959: chopping 4 bytes<br>08-23 14:49:43.724   803 20323 D iked    : [DEBUG]: isakmp.c:972: 80 bytes message received from 89.***.***.64[4500], rec = 0<br>08-23 14:49:43.724   803 20323 D iked    : [DEBUG]: ikev2.c:387: ikev2_input(0x7075331cf0, 0x7074dff268, 0x7074dff1e8)<br>08-23 14:49:43.724   803 20323 D iked    : [DEBUG]: ikev2.c:392: processing message version 2.000<br>08-23 14:49:43.724   803 20323 D iked    : [DEBUG]: ikev2_payload.c:114: ikev2_check_payloads(0x7075331cf0, 1)<br>08-23 14:49:43.724   803 20323 D iked    : [DEBUG]: ikev2_payload.c:145: offset 0x1c type 46 (ENCRYPTED) len 52<br>08-23 14:49:43.724   803 20323 I iked    : [INFO]: ikev2_payload.c:257: ikev2_donot_check_payload_before_decrypt, 46<br>08-23 14:49:43.724   803 20323 D iked    : [DEBUG]: ikev2.c:496: ike_sa: 0x7074e48800 ipsecid 0<br>08-23 14:49:43.725   803 20323 D iked    : [DEBUG]: ikev2_payload.c:735: ikev2_decrypt(0x7074e48800, 0x7075331cf0)<br>08-23 14:49:43.725   803 20323 D iked    : [DEBUG]: ikev2_payload.c:780: encryptor_decrypt begin<br>08-23 14:49:43.725   803 20323 I iked    : [INFO]: ikev2.c:579: payload info is not fragment payload, check payload<br>08-23 14:49:43.725   803 20323 D iked    : [DEBUG]: ikev2_payload.c:114: ikev2_check_payloads(0x7075331cf0, 0)<br>08-23 14:49:43.725   803 20323 D iked    : [DEBUG]: ikev2_payload.c:145: offset 0x1c type 46 (ENCRYPTED) len 4<br>08-23 14:49:43.725   803 20323 D iked    : [DEBUG]: ike_sa.c:1288: dpd polling interval 900<br>08-23 14:49:43.725   803 20323 D iked    : [DEBUG]: ikev2.c:630: dispatching ike_sa 0x7074e48800 state 7<br>08-23 14:49:43.725   803 20323 D iked    : [DEBUG]: ikev2.c:5594: request exch type 37<br>08-23 14:49:43.725   803 20323 D iked    : [DEBUG]: ikev2.c:855: update request message_id 0x3b6<br>08-23 14:49:43.725   803 20323 I iked    : [INFO]: ikev2.c:9820: pcscf_reselection_support is 0<br>08-23 14:49:43.725   803 20323 D iked    : [DEBUG]: ikev2_packet.c:322: ipsecid:0,ikev2_packet_construct(37, 0x28, 0x3b6, 0x7074e48800, [0x7074e15098, 0])<br>08-23 14:49:43.725   803 20323 D iked    : [DEBUG]: ikev2_payload.c:847: ikev2_encrypt(0x7074e48800, 0x7075331d00)<br>08-23 14:49:43.725   803 20323 D iked    : [DEBUG]: ikev2_payload.c:880: pad_len =15,block_len=16,n=0,max_padlen=-1<br>08-23 14:49:43.725   803 20323 D iked    : [DEBUG]: ikev2_payload.c:904: plaintext->l = 16,iv_len=16<br>08-23 14:49:43.725   803 20323 D iked    : [DEBUG]: ikev2_payload.c:965: result 0x7075331d10<br>08-23 14:49:43.725   803 20323 D iked    : [DEBUG]: ikev2_packet.c:453: result 0x7075331cb0,rec = 0<br>08-23 14:49:43.725   803 20323 D iked    : [DEBUG]: ikev2.c:904: ikev2_transmit_response(0x7074e48800, 0x7075331cb0) len 80<br>08-23 14:49:43.725   803 20323 D iked    : [DEBUG]: isakmp.c:2007: isakmp_transmit_noretry src: 192.***.***.24[4500] dst 89.***.***.64[4500] socket<br>08-23 14:49:43.725   803 20323 D iked    : [DEBUG]: sockmisc.c:368: send packet from 192.***.***.24[4500]<br>08-23 14:49:43.725   803 20323 D iked    : [DEBUG]: sockmisc.c:370: send packet to 89.***.***.64[4500]<br>08-23 14:49:43.725   803 20323 D iked    : [DEBUG]: sockmisc.c:547: update_card get_tos = 184<br>08-23 14:49:43.725   803 20323 D iked    : [DEBUG]: sockmisc.c:565: 1 times of 84 bytes message will be sent to 89.***.***.64[4500]<br>08-23 14:49:43.725   803 20323 I iked    : [INFO]: ikev2.c:9864: no available pcscf addr!!!<br>08-23 14:49:43.725   803 20323 I iked    : [INFO]: ike_localsocket.c:2882: Release WakeLock. vowifi_wake_lock_network<br>08-23 14:49:43.758  1262  1285 I ActivityManager: App 10224/com.termux.api targets O+, restricted<br>08-23 14:49:43.783  6836  6836 E termux-api: Error in TermuxApiReceiver<br>08-23 14:49:43.783  6836  6836 E termux-api: android.util.AndroidRuntimeException: Calling startActivity() from outside of an Activity  context requires the FLAG_ACTIVITY_NEW_TASK flag. Is this really what you want?<br>08-23 14:49:43.783  6836  6836 E termux-api:    at android.app.ContextImpl.startActivity(ContextImpl.java:1040)<br>08-23 14:49:43.783  6836  6836 E termux-api:    at android.app.ContextImpl.startActivity(ContextImpl.java:1009)<br>08-23 14:49:43.783  6836  6836 E termux-api:    at android.content.ContextWrapper.startActivity(ContextWrapper.java:379)<br>08-23 14:49:43.783  6836  6836 E termux-api:    at android.content.ContextWrapper.startActivity(ContextWrapper.java:379)<br>08-23 14:49:43.783  6836  6836 E termux-api:    at com.termux.api.TermuxApiReceiver.doWork(TermuxApiReceiver.java:122)<br>08-23 14:49:43.783  6836  6836 E termux-api:    at com.termux.api.TermuxApiReceiver.onReceive(TermuxApiReceiver.java:20)<br>08-23 14:49:43.783  6836  6836 E termux-api:    at android.app.ActivityThread.handleReceiver(ActivityThread.java:3996)<br>08-23 14:49:43.783  6836  6836 E termux-api:    at android.app.ActivityThread.access$2300(ActivityThread.java:273)<br>08-23 14:49:43.783  6836  6836 E termux-api:    at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2050)<br>08-23 14:49:43.783  6836  6836 E termux-api:    at android.os.Handler.dispatchMessage(Handler.java:112)<br>08-23 14:49:43.783  6836  6836 E termux-api:    at android.os.Looper.loop(Looper.java:216)<br>08-23 14:49:43.783  6836  6836 E termux-api:    at android.app.ActivityThread.main(ActivityThread.java:7625)<br>08-23 14:49:43.783  6836  6836 E termux-api:    at java.lang.reflect.Method.invoke(Native Method)<br>08-23 14:49:43.783  6836  6836 E termux-api:    at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:524)<br>08-23 14:49:43.783  6836  6836 E termux-api:    at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:987)<br>08-23 14:49:43.783  1262  4029 V BroadcastQueue: Finished with ordered broadcast BroadcastRecord{bacbffb u0 null}<br>08-23 14:49:43.785 12625 12625 D AndroidRuntime: Shutting down VM</pre></details>

**Comment by xalexalex** [comment]
The problem was in the request for permission. Fixed and will be published with the next release of termux-api. Meanwhile, you can manually grant the `read notifications` permission, and this should solve your problem.

**Comment by dsl400** [comment]
too late for the party ?
using android 5

`adb shell pm grant com.termux.api android.permission.BIND_NOTIFICATION_LISTENER_SERVICE`

`Operation not allowed: java.lang.SecurityException: Package com.termux.api has not requested permission android.permission.BIND_NOTIFICATION_LISTENER_SERVICE`
what am I doing wrong ?

**Comment by xalexalex** [comment]
@dsl400 I think that's unrelated to this issue; you should open a new one.

**Comment by dsl400** [comment]
> Meanwhile, you can manually grant the read notifications permission, 
in my humble opinion it is 100% related and somehow adds missing pieces to the puzzle
maybe there are other users that are not that experienced with manually granting permissions

**Comment by xalexalex** [comment]
Whoops, you're right. I meant it was unrelated to the issue (as in "the title of the issue"), but it is indeed related with my proposed solution.

However, termux-api does request that permission ([link](https://github.com/termux/termux-api/blob/master/app/src/main/AndroidManifest.xml#L78)), so I suspect it has something to do with either your termux-api not being the latest version or, more likely, some quirk with this permission in android 5.

What happens if you try to grant the permission not through `pm` but through the android settings?

**Comment by dsl400** [comment]
sadly emui 3.0 lite does not allow me to manually edit the permissions

**Comment by dsl400** [comment]
Not sure why but it works, my guess is that this huawei Y5 is to slow.
I typed termux-notification-list one more time just to see if it works and it hanged again but this time 
I just left the phone on the table and after more than a minute the phone was asking if I allow the app to access notifications. Prior to this I disabled then enabled notifications for termux and termux-api after reading [this](https://stackoverflow.com/a/40825389/2470661)

hope it helps!
